src/main/java/pl/javastart/streamsexercise/PaymentService.java

```diff
@@ -16,43 +19,84 @@
     }
 
     List<Payment> findPaymentsSortedByDateDesc() {
-        throw new RuntimeException("Not implemented");
+        return paymentRepository.findAll()
+                .stream()
+                .sorted(Comparator.comparing(Payment::getPaymentDate).reversed())
+                .collect(Collectors.toList());
     }
 
     List<Payment> findPaymentsForCurrentMonth() {
-        throw new RuntimeException("Not implemented");
+        return paymentRepository.findAll()
+                .stream()
+                .filter(payment -> Integer.valueOf(payment.getPaymentDate().getYear()).equals(dateTimeProvider.yearMonthNow().getYear()))
+                .filter(payment -> payment.getPaymentDate().getMonth().equals(dateTimeProvider.yearMonthNow().getMonth()))
+                .collect(Collectors.toList());
     }
 
     List<Payment> findPaymentsForGivenMonth(YearMonth yearMonth) {
-        throw new RuntimeException("Not implemented");
+        return paymentRepository.findAll()
+                .stream()
+                .filter(payment -> Integer.valueOf(payment.getPaymentDate().getYear()).equals(dateTimeProvider.yearMonthNow().getYear()))
+                .filter(payment -> payment.getPaymentDate().getMonth().equals(yearMonth.getMonth()))
+                .collect(Collectors.toList());
     }
 
     List<Payment> findPaymentsForGivenLastDays(int days) {
-        throw new RuntimeException("Not implemented");
+        ZonedDateTime thirtyDaysAgo = dateTimeProvider.zonedDateTimeNow().plusDays(-days);
+        return paymentRepository.findAll()
+                .stream()
+                .filter(payment -> payment.getPaymentDate().toInstant().isAfter(thirtyDaysAgo.toInstant()))
+                .collect(Collectors.toList());
     }
 
     Set<Payment> findPaymentsWithOnePaymentItem() {
-        throw new RuntimeException("Not implemented");
+        return paymentRepository.findAll()
+                .stream()
+                .filter(payment -> payment.getPaymentItems().size() == 1)
+                .collect(Collectors.toSet());
     }
 
     Set<String> findProductsSoldInCurrentMonth() {
-        throw new RuntimeException("Not implemented");
+        return findPaymentsForCurrentMonth().stream()
+                .flatMap(payment -> payment.getPaymentItems().stream())
+                .map(PaymentItem::getName)
+                .collect(Collectors.toSet());
     }
 
     BigDecimal sumTotalForGivenMonth(YearMonth yearMonth) {
-        throw new RuntimeException("Not implemented");
+        return findPaymentsForGivenMonth(yearMonth).stream()
+                .flatMap(payment -> payment.getPaymentItems().stream())
+                .map(PaymentItem::getFinalPrice)
+                .reduce(BigDecimal.ZERO, BigDecimal::add);
     }
 
     BigDecimal sumDiscountForGivenMonth(YearMonth yearMonth) {
-        throw new RuntimeException("Not implemented");
+        return findPaymentsForGivenMonth(yearMonth).stream()
```

Comment:
Tutaj można to było zrobić inaczej - prościej. Mogłeś wziąć wszystkie elementy z metody `findPaymentsForGivenMonth` następnie z nich za pomocą flatMap() wyciągnąć wszystkie paymentItems, następnie zrobić metodą `map` odjęcie finalnej ceny od regularnej, a na końcu redukcja. Nie musiałbyś robić tego ostatniego `subtract` z wywołaniem metody zliczającej ceny finalne.